File: src/api/derivatives.py
"""
Derivative tracking API blueprint.

This blueprint provides endpoints for intent evolution tracking:
- Get derivatives of an entry
- Get lineage/ancestry of an entry
- Get complete derivation tree
- Check derivative status
"""

from flask import Blueprint, jsonify, request

from . import state
from .utils import require_api_key, validate_json_schema

# Import derivative type constants
try:
    from blockchain import VALID_DERIVATIVE_TYPES
except ImportError:
    try:
        from src.blockchain import VALID_DERIVATIVE_TYPES
    except ImportError:
        VALID_DERIVATIVE_TYPES = {
            "amendment",
            "extension",
            "response",
            "revision",
            "reference",
            "fulfillment",
        }

# Create the blueprint
derivatives_bp = Blueprint("derivatives", __name__, url_prefix="/derivatives")


def _check_derivative_tracking():
    """Check if derivative tracking is enabled."""
    if not state.blockchain.enable_derivative_tracking:
        return jsonify(
            {
                "error": "Derivative tracking not enabled",
                "reason": "enable_derivative_tracking is False",
            }
        ), 503
    return None


def _validate_entry_ref(block_index: int, entry_index: int):
    """Validate that a block/entry reference exists."""
    if block_index < 0 or block_index >= len(state.blockchain.chain):
        return jsonify(
            {
                "error": "Block not found",
                "block_index": block_index,
                "valid_range": f"0-{len(state.blockchain.chain) - 1}",
            }
        ), 404

    block = state.blockchain.chain[block_index]
    if entry_index < 0 or entry_index >= len(block.entries):
        return jsonify(
            {
                "error": "Entry not found",
                "entry_index": entry_index,
                "valid_range": f"0-{len(block.entries) - 1}",
            }
        ), 404

    return None


@derivatives_bp.route("/types", methods=["GET"])
def get_derivative_types():
    """
    Get all valid derivative types.

    Returns:
        List of valid derivative type strings
    """
    return jsonify(
        {
            "types": list(VALID_DERIVATIVE_TYPES),
            "descriptions": {
                "amendment": "Modifies terms of parent entry",
                "extension": "Adds to parent without modifying",
                "response": "Response to parent entry",
                "revision": "Supersedes parent entirely",
                "reference": "Simply references parent",
                "fulfillment": "Fulfills/completes parent intent",
            },
        }
    )


@derivatives_bp.route("/<int:block_index>/<int:entry_index>", methods=["GET"])
def get_derivatives(block_index: int, entry_index: int):
    """
    Get all derivatives of a specific entry.

    Args:
        block_index: Block containing the parent entry
        entry_index: Index of the parent entry within the block

    Query params:
        recursive: If "true", get all descendants recursively (default: false)
        max_depth: Maximum recursion depth (default: 10)
        include_entries: If "true", include full entry data (default: false)

    Returns:
        List of derivative entry references
    """
    if error := _check_derivative_tracking():
        return error

    if error := _validate_entry_ref(block_index, entry_index):
        return error

    recursive = request.args.get("recursive", "false").lower() == "true"
    max_depth = request.args.get("max_depth", 10, type=int)
    include_entries = request.args.get("include_entries", "false").lower() == "true"

    # Bound max_depth to prevent abuse
    max_depth = min(max(1, max_depth), 50)

    result = state.blockchain.get_derivatives(
        block_index,
        entry_index,
        recursive=recursive,
        max_depth=max_depth,
        include_entries=include_entries,
    )

    return jsonify(result)


@derivatives_bp.route("/<int:block_index>/<int:entry_index>/lineage", methods=["GET"])
def get_lineage(block_index: int, entry_index: int):
    """
    Get the full ancestry/lineage of an entry.

    Traces back through all parent relationships to find
    the original root entries.

    Args:
        block_index: Block containing the entry
        entry_index: Index of the entry within the block

    Query params:
        max_depth: Maximum traversal depth (default: 10)
        include_entries: If "true", include full entry data (default: false)

    Returns:
        Lineage information including ancestors and roots
    """
    if error := _check_derivative_tracking():
        return error

    if error := _validate_entry_ref(block_index, entry_index):
        return error

    max_depth = request.args.get("max_depth", 10, type=int)
    include_entries = request.args.get("include_entries", "false").lower() == "true"

    # Bound max_depth
    max_depth = min(max(1, max_depth), 50)

    result = state.blockchain.get_lineage(
        block_index, entry_index, max_depth=max_depth, include_entries=include_entries
    )

    return jsonify(result)


@derivatives_bp.route("/<int:block_index>/<int:entry_index>/tree", methods=["GET"])
def get_derivation_tree(block_index: int, entry_index: int):
    """
    Get the complete derivation tree for an entry.

    Returns both ancestors and descendants in a tree structure.

    Args:
        block_index: Block containing the entry
        entry_index: Index of the entry within the block

    Query params:
        max_depth: Maximum traversal depth in each direction (default: 10)
        include_entries: If "true", include full entry data (default: false)

    Returns:
        Complete derivation tree with parents, lineage, roots, and derivatives
    """
    if error := _check_derivative_tracking():
        return error

    if error := _validate_entry_ref(block_index, entry_index):
        return error

    max_depth = request.args.get("max_depth", 10, type=int)
    include_entries = request.args.get("include_entries", "false").lower() == "true"

    # Bound max_depth
    max_depth = min(max(1, max_depth), 50)

    result = state.blockchain.get_derivation_tree(
        block_index, entry_index, max_depth=max_depth, include_entries=include_entries
    )

    return jsonify(result)


@derivatives_bp.route("/<int:block_index>/<int:entry_index>/status", methods=["GET"])
def get_derivative_status(block_index: int, entry_index: int):
    """
    Get derivative status for an entry.

    Returns whether the entry is a derivative and/or has derivatives.

    Args:
        block_index: Block containing the entry
        entry_index: Index of the entry within the block

    Returns:
        Derivative status information
    """
    if error := _check_derivative_tracking():
        return error

    if error := _validate_entry_ref(block_index, entry_index):
        return error

    is_derivative = state.blockchain.is_derivative(block_index, entry_index)
    has_derivatives = state.blockchain.has_derivatives(block_index, entry_index)

    # Get the entry to check for parent refs
    entry = state.blockchain.chain[block_index].entries[entry_index]

    result = {
        "block_index": block_index,
        "entry_index": entry_index,
        "is_derivative": is_derivative,
        "has_derivatives": has_derivatives,
        "derivative_type": entry.derivative_type if is_derivative else None,
        "parent_count": len(entry.parent_refs) if entry.parent_refs else 0,
    }

    if is_derivative and entry.parent_refs:
        result["parent_refs"] = entry.parent_refs

    return jsonify(result)


@derivatives_bp.route("/validate", methods=["POST"])
@require_api_key
def validate_derivative_refs():
    """
    Validate parent references before creating a derivative entry.

    Request body:
    {
        "parent_refs": [
            {"block_index": 1, "entry_index": 0},
            ...
        ],
        "derivative_type": "amendment"
    }

    Returns:
        Validation result with any issues found
    """
    if error := _check_derivative_tracking():
        return error

    data = request.get_json()

    if not data:
        return jsonify({"error": "No data provided"}), 400

    is_valid, error_msg = validate_json_schema(
        data, required_fields={"parent_refs": list, "derivative_type": str}, optional_fields={}
    )

    if not is_valid:
        return jsonify({"error": error_msg}), 400

    parent_refs = data["parent_refs"]
    derivative_type = data["derivative_type"]

    # Validate derivative type
    if derivative_type not in VALID_DERIVATIVE_TYPES:
        return jsonify(
            {
                "valid": False,
                "error": "Invalid derivative type",
                "derivative_type": derivative_type,
                "valid_types": list(VALID_DERIVATIVE_TYPES),
            }
        ), 400

    # Validate each parent reference
    issues = []
    valid_refs = []

    for i, ref in enumerate(parent_refs):
        block_idx = ref.get("block_index")
        entry_idx = ref.get("entry_index")

        if block_idx is None or entry_idx is None:
            issues.append({"ref_index": i, "error": "Missing block_index or entry_index"})
            continue

        if block_idx < 0 or block_idx >= len(state.blockchain.chain):
            issues.append({"ref_index": i, "block_index": block_idx, "error": "Block not found"})
            continue

        block = state.blockchain.chain[block_idx]
        if entry_idx < 0 or entry_idx >= len(block.entries):
            issues.append({"ref_index": i, "entry_index": entry_idx, "error": "Entry not found"})
            continue

        # Valid reference
        entry = block.entries[entry_idx]
        valid_refs.append(
            {
                "block_index": block_idx,
                "entry_index": entry_idx,
                "author": entry.author,
                "intent": entry.intent,
            }
        )

    return jsonify(
        {
            "valid": len(issues) == 0,
            "derivative_type": derivative_type,
            "total_refs": len(parent_refs),
            "valid_refs": valid_refs,
            "issues": issues,
        }
    )

File: _deferred/sdk/src/clients/derivatives.ts
/**
 * Derivatives Client - Intent Evolution Tracking
 *
 * Handles derivative/lineage operations:
 * - Get derivatives of an entry
 * - Get lineage/ancestry of an entry
 * - Get complete derivation tree
 * - Validate derivative references
 */

import { HttpClient } from '../utils/client';
import type { Entry } from '../types';

/** Valid derivative types */
export type DerivativeType =
  | 'amendment'
  | 'extension'
  | 'response'
  | 'revision'
  | 'reference'
  | 'fulfillment';

/** Parent reference for derivative entries */
export interface ParentRef {
  block_index: number;
  entry_index: number;
  relationship?: string;
}

/** Derivative entry reference */
export interface DerivativeRef {
  block_index: number;
  entry_index: number;
  derivative_type: DerivativeType;
  relationship: string;
  registered_at: number;
  depth?: number;
  entry?: Entry;
}

/** Lineage/ancestor reference */
export interface LineageRef {
  block_index: number;
  entry_index: number;
  relationship: string;
  depth?: number;
  entry?: Entry;
}

/** Derivatives response */
export interface DerivativesResponse {
  parent: {
    block_index: number;
    entry_index: number;
  };
  derivative_count: number;
  recursive: boolean;
  derivatives: DerivativeRef[];
}

/** Lineage response */
export interface LineageResponse {
  entry: {
    block_index: number;
    entry_index: number;
  };
  ancestor_count: number;
  lineage: LineageRef[];
  roots: Array<{
    block_index: number;
    entry_index: number;
    entry?: Entry;
  }>;
}

/** Derivation tree response */
export interface DerivationTreeResponse {
  entry: {
    block_index: number;
    entry_index: number;
    data?: Entry;
  };
  parents: LineageRef[];
  lineage: LineageRef[];
  roots: Array<{
    block_index: number;
    entry_index: number;
    entry?: Entry;
  }>;
  derivatives: DerivativeRef[];
  all_descendants: DerivativeRef[];
}

/** Derivative status response */
export interface DerivativeStatusResponse {
  block_index: number;
  entry_index: number;
  is_derivative: boolean;
  has_derivatives: boolean;
  derivative_type: DerivativeType | null;
  parent_count: number;
  parent_refs?: ParentRef[];
}

/** Derivative types response */
export interface DerivativeTypesResponse {
  types: DerivativeType[];
  descriptions: Record<DerivativeType, string>;
}

/** Validate derivative refs request */
export interface ValidateDerivativeRefsRequest {
  parent_refs: ParentRef[];
  derivative_type: DerivativeType;
}

/** Validate derivative refs response */
export interface ValidateDerivativeRefsResponse {
  valid: boolean;
  derivative_type: DerivativeType;
  total_refs: number;
  valid_refs: Array<{
    block_index: number;
    entry_index: number;
    author: string;
    intent: string;
  }>;
  issues: Array<{
    ref_index: number;
    error: string;
    block_index?: number;
    entry_index?: number;
  }>;
}

/** Options for getting derivatives */
export interface GetDerivativesOptions {
  /** If true, get all descendants recursively */
  recursive?: boolean;
  /** Maximum recursion depth (default: 10, max: 50) */
  maxDepth?: number;
  /** If true, include full entry data */
  includeEntries?: boolean;
}

/** Options for getting lineage */
export interface GetLineageOptions {
  /** Maximum traversal depth (default: 10, max: 50) */
  maxDepth?: number;
  /** If true, include full entry data */
  includeEntries?: boolean;
}

/**
 * Derivatives client for intent evolution tracking
 */
export class DerivativesClient {
  constructor(private readonly http: HttpClient) {}

  /**
   * Get all valid derivative types and their descriptions
   *
   * @returns List of valid derivative types
   *
   * @example
   * ```ts
   * const types = await client.derivatives.getTypes();
   * console.log(types.types); // ['amendment', 'extension', ...]
   * console.log(types.descriptions.amendment); // 'Modifies terms of parent entry'
   * ```
   */
  async getTypes(): Promise<DerivativeTypesResponse> {
    return this.http.get<DerivativeTypesResponse>('/derivatives/types');
  }

  /**
   * Get all derivatives of a specific entry
   *
   * @param blockIndex - Block containing the parent entry
   * @param entryIndex - Index of the parent entry within the block
   * @param options - Query options
   * @returns List of derivative entry references
   *
   * @example
   * ```ts
   * // Get direct derivatives only
   * const derivatives = await client.derivatives.getDerivatives(1, 0);
   *
   * // Get all descendants recursively with full entry data
   * const allDescendants = await client.derivatives.getDerivatives(1, 0, {
   *   recursive: true,
   *   maxDepth: 5,
   *   includeEntries: true
   * });
   *
   * for (const deriv of allDescendants.derivatives) {
   *   console.log(`${deriv.derivative_type} at depth ${deriv.depth}`);
   * }
   * ```
   */
  async getDerivatives(
    blockIndex: number,
    entryIndex: number,
    options?: GetDerivativesOptions
  ): Promise<DerivativesResponse> {
    const params: Record<string, string | number> = {};

    if (options?.recursive) params.recursive = 'true';
    if (options?.maxDepth) params.max_depth = options.maxDepth;
    if (options?.includeEntries) params.include_entries = 'true';

    return this.http.get<DerivativesResponse>(
      `/derivatives/${blockIndex}/${entryIndex}`,
      params
    );
  }

  /**
   * Get the full ancestry/lineage of an entry
   *
   * Traces back through all parent relationships to find
   * the original root entries.
   *
   * @param blockIndex - Block containing the entry
   * @param entryIndex - Index of the entry within the block
   * @param options - Query options
   * @returns Lineage information including ancestors and roots
   *
   * @example
   * ```ts
   * const lineage = await client.derivatives.getLineage(5, 2, {
   *   includeEntries: true
   * });
   *
   * console.log(`Entry has ${lineage.ancestor_count} ancestors`);
   * console.log('Root entries:', lineage.roots);
   *
   * for (const ancestor of lineage.lineage) {
   *   console.log(`Ancestor at depth ${ancestor.depth}: ${ancestor.entry?.intent}`);
   * }
   * ```
   */
  async getLineage(
    blockIndex: number,
    entryIndex: number,
    options?: GetLineageOptions
  ): Promise<LineageResponse> {
    const params: Record<string, string | number> = {};

    if (options?.maxDepth) params.max_depth = options.maxDepth;
    if (options?.includeEntries) params.include_entries = 'true';

    return this.http.get<LineageResponse>(
      `/derivatives/${blockIndex}/${entryIndex}/lineage`,
      params
    );
  }

  /**
   * Get the complete derivation tree for an entry
   *
   * Returns both ancestors and descendants in a tree structure.
   *
   * @param blockIndex - Block containing the entry
   * @param entryIndex - Index of the entry within the block
   * @param options - Query options
   * @returns Complete derivation tree
   *
   * @example
   * ```ts
   * const tree = await client.derivatives.getTree(3, 1, {
   *   maxDepth: 10,
   *   includeEntries: true
   * });
   *
   * console.log('Parents:', tree.parents.length);
   * console.log('Full lineage:', tree.lineage.length);
   * console.log('Roots:', tree.roots.length);
   * console.log('Direct derivatives:', tree.derivatives.length);
   * console.log('All descendants:', tree.all_descendants.length);
   * ```
   */
  async getTree(
    blockIndex: number,
    entryIndex: number,
    options?: GetLineageOptions
  ): Promise<DerivationTreeResponse> {
    const params: Record<string, string | number> = {};

    if (options?.maxDepth) params.max_depth = options.maxDepth;
    if (options?.includeEntries) params.include_entries = 'true';

    return this.http.get<DerivationTreeResponse>(
      `/derivatives/${blockIndex}/${entryIndex}/tree`,
      params
    );
  }

  /**
   * Get derivative status for an entry
   *
   * Returns whether the entry is a derivative and/or has derivatives.
   *
   * @param blockIndex - Block containing the entry
   * @param entryIndex - Index of the entry within the block
   * @returns Derivative status information
   *
   * @example
   * ```ts
   * const status = await client.derivatives.getStatus(2, 0);
   *
   * if (status.is_derivative) {
   *   console.log(`This is a ${status.derivative_type} of ${status.parent_count} parent(s)`);
   * }
   *
   * if (status.has_derivatives) {
   *   console.log('This entry has been derived from');
   * }
   * ```
   */
  async getStatus(
    blockIndex: number,
    entryIndex: number
  ): Promise<DerivativeStatusResponse> {
    return this.http.get<DerivativeStatusResponse>(
      `/derivatives/${blockIndex}/${entryIndex}/status`
    );
  }

  /**
   * Validate parent references before creating a derivative entry
   *
   * Use this to check that parent references are valid before
   * submitting a derivative entry.
   *
   * @param request - Validation request with parent refs and derivative type
   * @returns Validation result with any issues found
   *
   * @example
   * ```ts
   * const validation = await client.derivatives.validateRefs({
   *   parent_refs: [
   *     { block_index: 1, entry_index: 0 },
   *     { block_index: 2, entry_index: 1 }
   *   ],
   *   derivative_type: 'amendment'
   * });
   *
   * if (validation.valid) {
   *   // Safe to create derivative entry
   *   await client.core.createEntry({
   *     content: 'Amending the terms from Block 1, Entry 0...',
   *     author: 'user@example.com',
   *     intent: 'Amendment to original agreement',
   *     metadata: {
   *       parent_refs: validation.valid_refs,
   *       derivative_type: 'amendment'
   *     }
   *   });
   * } else {
   *   console.error('Invalid parent refs:', validation.issues);
   * }
   * ```
   */
  async validateRefs(
    request: ValidateDerivativeRefsRequest
  ): Promise<ValidateDerivativeRefsResponse> {
    return this.http.post<ValidateDerivativeRefsResponse>(
      '/derivatives/validate',
      request
    );
  }

  /**
   * Check if an entry is a derivative (has parents)
   *
   * Convenience method that wraps getStatus.
   *
   * @param blockIndex - Block containing the entry
   * @param entryIndex - Index of the entry within the block
   * @returns True if the entry is a derivative
   */
  async isDerivative(blockIndex: number, entryIndex: number): Promise<boolean> {
    const status = await this.getStatus(blockIndex, entryIndex);
    return status.is_derivative;
  }

  /**
   * Check if an entry has derivatives (has children)
   *
   * Convenience method that wraps getStatus.
   *
   * @param blockIndex - Block containing the entry
   * @param entryIndex - Index of the entry within the block
   * @returns True if the entry has derivatives
   */
  async hasDerivatives(blockIndex: number, entryIndex: number): Promise<boolean> {
    const status = await this.getStatus(blockIndex, entryIndex);
    return status.has_derivatives;
  }
}
